src/common/util/queueImmut.ml

```diff
@@ -0,0 +1,126 @@
+(* A Queue (FIFO, imutable) implementation, using two Lists for efficient push and pop handling.
+
+   Example Queue: remove here, 1, 2, 3, 4, add here
+   Representation as Queue: Queue([1,2], [4,3]) or Queue([1], [4,3,2]) or Queue([1,2,3,4], []) or ...
+*)
+
+exception Empty
+
+type 'a t = Queue of 'a list * 'a list
```

Comment:
Might be worth considering adding an `*.mli` here, and not exposing the type to the outside, then in here, you can remove the `Queue` constructor and just go with having the type be two lists. That would mean it's a bit less code in the end.